rfcs/0108-nixos-containers.md

```diff
@@ -0,0 +1,460 @@
+
+```nix
+{
+  nixos.containers.zones.demo = {};
+  nixos.containers.instances = {
+    test1.network.zone = "demo";
+    test2.network.zone = "demo";
+  };
+}
+```
+
+IP addresses can be statically assigned to a container as well:
+
+``` nix
+{
+  nixos.containers.instances.static = {
+    network = {
+      v4.static.containerPool = [ "10.237.1.3/16" ];
+      v6.static.containerPool = [ "2a01:4f9:4b:1659:3aa3:cafe::3/96" ];
+    };
+    config = {};
+  };
+}
+```
+
+With this change, the containers live in the given subnets and both on the host- and container-side
+the network will be properly configured accordingly.
+
+### Advanced Features
+
+MACVLANs are an example for how every unit setting from `networkd` and `nspawn` can be used.
+These are helpful to assign multiple virtual interfaces with distinct MAC addresses to a single
+physical NIC.
+
+A sub-interface which is actually part of the physical one can be moved into the container's
+namespace then:
+
+``` nix
+{
+  # Config for the physical interface itself with DHCP enabled and associated to a MACVLAN.
+  systemd.network.networks."40-eth1" = {
+    matchConfig.Name = "eth1";
+    networkConfig.DHCP = "yes";
+    dhcpConfig.UseDNS = "no";
+    networkConfig.MACVLAN = "mv-eth1-host";
+    linkConfig.RequiredForOnline = "no";
+    address = lib.mkForce [];
+    addresses = lib.mkForce [];
+  };
+
+  # The host-side sub-interface of the MACVLAN. This means that the host is reachable
+  # within the (internal) network at `192.168.2.2`.
+  systemd.network.networks."20-mv-eth1-host" = {
+    matchConfig.Name = "mv-eth1-host";
+    networkConfig.IPForward = "yes";
+    dhcpV4Config.ClientIdentifier = "mac";
+    address = lib.mkForce [
+      "192.168.2.2/24"
+    ];
+  };
+  systemd.network.netdevs."20-mv-eth1-host" = {
+    netdevConfig = {
+      Name = "mv-eth1-host";
+      Kind = "macvlan";
+    };
+    extraConfig = ''
+      [MACVLAN]
+      Mode=bridge
+    '';
+  };
+
+  # Assign a MACVLAN to a container. This is done by pure nspawn.
+  systemd.nspawn.vlandemo.networkConfig.MACVLAN = "eth1";
+  nixos.containers = {
+    instances.vlandemo.config = {
+      systemd.network = {
+        networks."10-mv-eth1" = {
+          matchConfig.Name = "mv-eth1";
+          address = [ "192.168.2.5/24" ];
+        };
+        netdevs."10-mv-eth1" = {
+          netdevConfig.Name = "mv-eth1";
+          netdevConfig.Kind = "veth";
+        };
+      };
+    };
+  };
+}
+```
+
+### Imperative containers
+
+#### Create a container with a pinned `nixpkgs`
+
+Let the following expression be called `imperative-container.nix`:
+
+```nix
+{
+  config.nixpkgs = <nixpkgs>;
+  config.config = { pkgs, ... }: {
+    services.nginx.enable = true;
+    networking.firewall.allowedTCPPorts = [ 80 ];
+  };
+
+  # This implies that the "default" networking mode (i.e. DHCPv4) is used
+  # and not the host's network (which is the default for imperative containers).
+  config.network = {};
```

Comment:
Maybe this should be called `container.network` to avoid confusion with the `networking` namespace available in all NixOS configurations.

Reply:
The primary issue is something different: I think that naming the container's config-attribute `config` was a poor choice actually because we now have to do some nasty hacks such as `config.config` (and thus also `config.network` which is indeed confusing).

I'll change it to let's say `system-config`, then the file would look like this:

``` nix
{
  nixpkgs = <nixpkgs>;
  network = { ... };
  system-config = { pkgs, ... }: {
    # ...
    networking.firewall.allowedTCPPorts = [ 80 ];
  };
}
```